common/environment/environment.cpp

```diff
@@ -1606,27 +1646,77 @@ IConstDropZoneInfo * CLocalEnvironment::getDropZoneByAddressPath(const char * ne
-        // dz1: /a/b/c/d
-        // dz2: /a/b/c
-        // dz3: /a/b
-        // check all and select the shortest
-
         if (strncmp(pdropzoneDir, targetFilePath, strlen(pdropzoneDir)) == 0)
         {
-            if (dropzonePathLen > strlen(pdropzoneDir))
+            dropZone = &zoneIt->query();
+
+            // For backward compatibility
+            SCMStringBuffer dropZoneMachineName;
+            // Returns dropzone '@computer' value if it is exists
+            dropZone->getComputerName(dropZoneMachineName);
+            if ( 0 != dropZoneMachineName.length())
             {
-                dropzonePathLen = strlen(pdropzoneDir);
-                dropzonePath.set(pdropzoneDir);
-                pdropzonePath = dropzonePath.str();
-                retVal = &zoneIt->query();
+                Owned<IConstMachineInfo> machineInfo = getMachine(dropZoneMachineName.str());
+                if (machineInfo)
+                {
+                    SCMStringBuffer dropZoneMachineNetAddress;
+                    machineInfo->getNetAddress(dropZoneMachineNetAddress);
+                    IpAddress dropZoneMachineIp(dropZoneMachineNetAddress.str());
+#ifdef _DEBUG
+                    StringBuffer dropZoneMachineNetAddressIpString;
+                    dropZoneMachineIp.getIpText(dropZoneMachineNetAddressIpString);
+                    LOG(MCdebugInfo, unknownJob, "Drop zone coputer: '%s', IP: '%s'", dropZoneMachineName.str(), dropZoneMachineNetAddressIpString.str());
+#endif
+                    if (targetIp.ipequals(dropZoneMachineIp))
+                       break;
+                    else
+                       // Path matched, but machine didn't
+                       dropZone = nullptr;
+                }
+
+            }
+
+            // check if it is local machine
+            //if (strcmp(netaddress, dropZoneMachineName.str()) !=  0)
+            else
+            {
+                // It isn't local machine or it has not '@server' attrib. Check the ServerList
+                bool isServerFound = false;
+
+                Owned<IConstDropZoneServerInfoIterator> dropzoneServerListIt = dropZone->getServers();
+                ForEach(*dropzoneServerListIt)
+                {
+                    StringBuffer dropzoneServer;
+                    dropzoneServerListIt->query().getServer(dropzoneServer);
+                    // It can be a hostname or an IP -> get the IP
+                    IpAddress serverIP(dropzoneServer.str());
+
+#ifdef _DEBUG
+                    StringBuffer serverIpString;
+                    serverIP.getIpText(serverIpString);
+                    LOG(MCdebugInfo, unknownJob, "Listed server: '%s', IP: '%s'", dropzoneServer.str(), serverIpString.str());
+#endif
+                    if (targetIp.ipequals(serverIP))
+                    {
+                        // OK the target is a valid machine on the white list
+                        isServerFound = true;
+                        break;
```

Comment:
see comment above, think it should still continue, remembering current match path length, for backward compatibility reasons.